python4j/python4j-core/src/main/java/org/eclipse/python4j/PythonGIL.java

```diff
@@ -0,0 +1,86 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Konduit K.K.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License, Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ******************************************************************************/
+
+package org.eclipse.python4j;
+
+
+import org.bytedeco.cpython.PyThreadState;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.bytedeco.cpython.global.python.*;
+
+
+public class PythonGIL implements AutoCloseable {
+    private static PyThreadState mainThreadState;
+    private static final AtomicBoolean acquired = new AtomicBoolean();
+    private boolean acquiredByMe = false;
+    private static long defaultThreadId = -1;
+
+    public  static void assertThreadSafe(){
+        if (acquired.get()){
+            return;
+        }
+        if (defaultThreadId == -1){
+            defaultThreadId = Thread.currentThread().getId();
+        }
+        else if (defaultThreadId != Thread.currentThread().getId()){
+            throw new RuntimeException("Attempt to use Python4j from multiple threads without " +
+                    "acquiring GIL. Enclose your code in a try(PythonGIL gil = PythonGIL.lock()){...}" +
+                    " block to ensure that GIL is acquired in multi-threaded environments.");
+        }
+
+
+    }
+
+    static {
+        mainThreadState = PyThreadState_Get();
+    }
+
+    private PythonGIL() {
+        if (!acquired.get()){
+            acquire();
+            acquiredByMe = true;
+            acquired.set(true);
+        }
+    }
+
+    @Override
+    public void close() {
+        if (acquiredByMe){
+            release();
+            acquired.set(false);
+            acquiredByMe = false;
+        }
+
+    }
```

Comment:
close method should probably also be synchronized

Reply:
(Multi-thread tests added)